### ISSUE TITLE
React-to-print with Electron silent printing

### ISSUE THREAD

**Issue by ririshi** [MatthewHerbst/react-to-print#270]
Our dev team is currently working on a react dashboard that receives receipts and automatically prints them. We would like this to work with Electron, but its support for silent printing differs significantly from how normal web browsers handle this.

When Google Chrome is used in Kiosk printing mode (experimental flag in Chrome, also available under other names for other browsers), `window.print()` will silently print using the default printer. Electron also supports silent printing, but requires the rendering process to send an ipc message to the Electron main process ([docs](https://www.electronjs.org/docs/api/ipc-main)), which can in turn call the `webContents.print()` command with `{ silent: true }` in its `options` argument ([docs](https://www.electronjs.org/docs/api/web-contents#contentsprintoptions-callback)).

We used the Electron `contextBridge` ([docs](https://www.electronjs.org/docs/api/context-bridge)) to expose a small "API" with `window.electron.printSilently()` that will send an ipc message from the rendering process to the Electron main process, which will print the contents of the sender process' window. In order to make this work with react-to-print, we would probably have to change the source code and replace the normal `window.print()` call with a `window.electron.printSilently()` call.

Our dev team doesn't have any prior experience with modifying npm packages, so any help would be much appreciated. I don't think this use case is broad enough to warrant a separate package on npm, but perhaps the code here could be changed to allow a callback instead of only calling the default `window.print()` on the iframe. That would make this package more flexible by allowing users to "insert" their own print handlers.

The caveat here is that the callback may not be able to invoke methods on the iframe window, but I'm not sure how context is handled if a callback is passed. It would have to call the iframe's `window.electron.printSilently()` in our case.

Interested to see your ideas and suggestions!

**Comment by MatthewHerbst** [comment]
Hi there. That is a very interesting use case, thanks for bringing this up. It certainly would be trivial to allow the calling user to pass a callback that is used instead of `target.contentWindow.print()`. `target` is the `HTMLIFrameElement`, so that could also easily be passed to the callback function to do with as it wishes, at which point I think the callback function should have full access to the iframe since we don't sandbox the iframe in any way.

> Our dev team doesn't have any prior experience with modifying npm packages

Nothing special about npm here. We just have to modify the code here on GitHub, and then I push a version to npm

Check out #272 which I just pushed up, and let me know if that might work for you. I would ask before merging that that you provide me with sample code I can use to show how the new method could work

**Comment by ririshi** [comment]
Wow, thanks for the quick reply. Sorry it's taking me longer to respond. I think that callback option should work fine for us.

What kind of sample code would you like to see? Something like what we use for Electron, or just a generic example? And should I post it on this thread?

This is probably the first time I'm "contributing" to an open source project, so I'm still finding my footing.

**Comment by MatthewHerbst** [comment]
No problem, and happy to walk you through it. The best way would be if you did a PR directly into my existing branch. So, follow these steps:

```sh
git clone git@github.com:MatthewHerbst/react-to-print.git
cd react-to-print
git checkout print-replacer
git checkout -B print-replacer-example
```

Then, add an example file or directory in the `example` folder. Ideally it would be complete enough to show anyone else using Electron how to do what you are doing. Maybe call the example `ElectronPrint` or something. Don't worry about adding it to the `index.tsx` file in the directory. Once you've add it, just make a commit and open a PR against my branch:

```sh
git add .
git commit
git push origin print-replacer-example
```

That last command should give you a link in the console that you can follow to open a new Pull Request. You'll want to target merging your branch (`ririshi/print-replacer-example`) into my branch (`matthewherbst/print-replacer`)

Let me know if you have any trouble, or need any help with the git commands

**Comment by ririshi** [comment]
Thanks for running me through the steps. I think you meant `grgnb/react-to-print` instead of your own user? Will add a slimmed-down version our electron `main.js` and `electron_preloader.js` to show the Electron part of things, and make a simple React component to actually show the printing. We're not using typescript at all, though, is that okay?

Edit: I just realized I am beinga bit stupid. I can happily write all sorts of example code, but I should first test whether the solution even works for us at all. Could you explain how I can use the new branch of the package in my project, as it won't be going through npm and thus won't be automatically available by simply importing react-to-print?

**Comment by ririshi** [comment]
Found an issue on the Electron side of things while testing the callback. When sending an IPC message to the Electron main process, it does not know how to differentiate between the main browser window and the IPC message sender. This results in the print callback always printing the whole window, instead of just the iframe. Not sure how to tackle it. Tried to find the Electron Slack to ask an expert, but had no luck.

Edit: Pushed my initial draft to a fork [here](https://github.com/Ririshi/react-to-print/tree/print-replacer-example). I think Electron's inability to select a specific iframe's `webContents` is a breaking issue for this idea, but I hope there is some way to specifically call the `print` method on an iframe's contents.

**Comment by MatthewHerbst** [comment]
I'm glad you found a way to test it. I can't say I've ever used Electron before, so I'm not going to be of much help figuring that out. Maybe if you post a question on StackOverflow about it you could get a good answer? Then we can incorporate that into the PR I have if there is any additional information needed by the callback

**Comment by ririshi** [comment]
With some creative thinking, I made a workaround (https://github.com/Ririshi/react-to-print/commit/d3e5c66726e233626b4745740da6f0a772e5d417). It's rather ugly, but it works. Electron is not able to retrieve the `webContents` instance for iframes, so I created a new window, which sends the printing command to Electron, gets printed, and is subsequently closed.

It works fine now, but it would be nice to have additional error handling abilities. For example, raising an error in the print function simply prints a console message saying something went wrong. It would be nice to be able to handle the error in the `onPrintError` handler. Added a small sample in the code with a `TODO` tag to highlight it.

All in all not the final solution I was hoping for, since react-to-print now creates an iframe that is not really used at all, but I guess it will have to do, unless I can find a better option somehow.

Edit: This is the first time I'm writing TypeScript ever in my life, so there are probably a few things that aren't done correctly. Just a heads-up.

**Comment by ririshi** [comment]
On a related note, I encountered some issues while committing my new code. Perhaps this warrants a separate issue, but while trying to add the example code to the project, the Husky pre-commit hook fails with this message:

```
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: example\ElectronPrint\index.tsx.
The file must be included in at least one of the projects provided
```

It's peculiar that it only highlights this item, since it looks like the `tsconfig.json` file does not include any of the `example` files, so it should either be throwing this error for every file in the directory, or none at all. I don't want to mess up the webpack and typescript configs, so I haven't changed anything about it in my PR, instead using the `--no-verify` flag, since my IDE already highlighted all the TypeScript errors and I've resolved any errors it showed before trying to commit.

**Comment by ririshi** [comment]
Is there a chance you could follow up on both of our PRs, @MatthewHerbst? Would love to be able to use the callback functionality in our project soon.

**Comment by MatthewHerbst** [comment]
Hey, apologies, I recently moved and had to take some time off from coding. I will add some error handling with `onPrintError`, that's good feedback. I'll merge it and get an RC version ready in a little bit

**Comment by MatthewHerbst** [comment]
I don't think we can do anything for now about not creating the iframe. That's very fundamental to how the code currently works

**Comment by MatthewHerbst** [comment]
Should be available in `v2.10.0` let me know how it works for you!

**Comment by MatthewHerbst** [comment]
Working Electron example with how to print using `react-to-print` is now available! https://stackoverflow.com/questions/68751490/electron-print-iframe-given-reference-to-it
